Concept + endpoint documentation for the Geostore service

edwardpmorris opened the pull request:
Documented v1 and v2 geostore endpoints. A few content and style points to chat about, and a few end points I was unable to workout how to use:

+ POST geostore using ESRI JSON (seems to be possible in the code, but details unclear) 
+ POST geostore/area (Not sure what this does or how to use it)
+ GET geostore/admin/iso/id1/id2 (No way of getting id2?)  

One general comment about the micro-service is that the admin, wdpa, and use endpoints are a bit "application" specific and opinionated (as in the sources of geometries, and whether or not simplification and filtering of the data source is a applied). Also it was unclear to me what happens if you create a geoStore using a  GeoJSON FeatureCollection with multiple features.

tiagojsag commented on source/includes/api/_geostore.md:
```diff
@@ -1,58 +1,71 @@
 # Geostore
 
-## What is Geostore?
+***Note — This is the documentation for version 1 of the Geostore microservice, the latest version is [version 2](https://resource-watch.github.io/doc-api/index-rw.html#geoStore-v2).***
```
Right now the differences between the two files seem to be minimal, and AFAIK functionally they are too. Any chance we can merge the two files and DRY things?

edwardpmorris replied:
Removed.

tiagojsag commented on source/includes/api/_geostore.md:
```diff
@@ -1,58 +1,71 @@
-Geostore object contains the following fields:
+Geostore is a database for storing geoStore objects which contain geographic data structures described using the [GeoJSON](https://geojson.org/) schema. GeoJSON supports the following geometry types: `Point`, `LineString`, `Polygon`, `MultiPoint`, `MultiLineString`, and `MultiPolygon`. Geometric objects with additional properties are `Feature` objects. Sets of features are contained by `FeatureCo ...
 
+The Geostore microservice provides a comprehensive way to communicate geographic data structures between platforms and microservices. It enables users to [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) geographic data structures from the Geostore, and query information about a geoStore object, such as the bounding box of all its' geometries.
```
I get the idea between using the CRUD term, but I don't think it applies here - I think you cant delete geostores, and updates are limited AFAIK. I think this may ultimately be misleading to readers

edwardpmorris replied:
Changed this section to be simpler and more readable. Removed reference to CRUD.

tiagojsag commented on source/includes/api/_geostore.md:
```diff
@@ -1,58 +1,71 @@
-<aside class="notice">
-Remember — All endpoint of geostore don't need that you are authenticated.
-</aside>
+After viewing the documentation below, consider looking at the Geostore tutorial for a step-by-step guide to rendering an actual geoStore object on a web application.
```
this should have a link to the corresponding section

edwardpmorris replied:
Removed this reference. We can add a link to tutorial when one is made.

tiagojsag commented on source/includes/api/_geostore.md:
```diff
@@ -1,58 +1,71 @@
+After viewing the documentation below, consider looking at the Geostore tutorial for a step-by-step guide to rendering an actual geoStore object on a web application.
 
-## Create Geostore
+***Remember — All of the Geostore endpoints do not require that users are authenticated.***
```
remove this please, let's not document things the user _doesn't_ need

edwardpmorris replied:
Removed.

tiagojsag commented on source/includes/api/_geostore.md:
```diff
@@ -1,58 +1,71 @@
+## What is Geostore
 
-Geostore object contains the following fields:
+Geostore is a database for storing geoStore objects which contain geographic data structures described using the [GeoJSON](https://geojson.org/) schema. GeoJSON supports the following geometry types: `Point`, `LineString`, `Polygon`, `MultiPoint`, `MultiLineString`, and `MultiPolygon`. Geometric objects with additional properties are `Feature` objects. Sets of features are contained by `FeatureCo ...
```
I'd suggest something like: 


`A geostore is an object which contain geographic data structures described using the GeoJSON schema. If you are new to GeoJSON, you can read a quick description on its [Wikipedia page](https://en.wikipedia.org/wiki/GeoJSON), or you can read the full specification on its [official page](https://geojson.org/). `

The official page only has the spec, nobody will read that :P

edwardpmorris replied:
Moved this information to concept/_geostore.md. In my opinion, from a GIS user perspective, understanding something about the specifics of GeoJSON and how this relates to the design considerations of Geostore is important. I gave 3 options of varying complexity; overview article focused on developers, links to JSON schema, and the full spec. Also highlighted the important things to consider. see concept/_geostore.md

edwardpmorris commented:
Outstanding questions:

+ confirm Geostore action when adding exactly the same GeoJSON (includes provider definition), which should create exactly the same Geostore ID and hash?
+ Cite/link to source(s) of Land-use types data.
+ The WDPA is potentially updated monthly. Specify which version is supplied by Geostore and what the update policy is. Note protected planet have an API for accessing the data set.  
+ Find out about and document ["Create using ESRI JSON"](https://github.com/gfw-api/gfw-geostore-api/blob/ab23215e9887611f60b2662eae3e107142c5fa01/app/src/routes/api/v2/geostore.router.js#L77)

edwardpmorris commented:
This call appears to return an error:

``` bash
curl --location --request POST 'https://api.resourcewatch.org/v1/geostore/area' \
--header 'Content-Type: application/json' \
--data-raw '{
   "geojson":{
      "type":"FeatureCollection",
      "features":[
         {
            "type":"Feature",
            "properties":{

            },
            "geometry":{
               "type":"Polygon",
               "coordinates":[
                  [
                     [
                        -4.4549560546875,
                        40.84706035607122
                     ],
                     [
                        -4.4549560546875,
                        41.30257109430557
                     ],
                     [
                        -3.5211181640624996,
                        41.30257109430557
                     ],
                     [
                        -3.5211181640624996,
                        40.84706035607122
                     ],
                     [
                        -4.4549560546875,
                        40.84706035607122
                     ]
                  ]
               ]
            }
         }
      ]
   }
}'
`` 

``
{
    "errors": [
        {
            "status": 404,
            "detail": "GeoJSONConverter.convert is not a function"
        }
    ]
}
```

edwardpmorris commented:
This call gives and authorization error, which is unexpected. Expected ALL geostore end points to NOT require authorization. 

```bash
curl --location --request GET 'https://api.resourcewatch.org/v2/geostore/35a6d982388ee5c4e141c2bceac3fb72/view'
```
```
{
    "errors": [
        {
            "status": 401,
            "detail": "Requires authentication"
        }
    ]
}
```

tiagojsag reviewed (changes requested):
- I really liked the rewritten intro texts
- There are multiple typos, please use a spellchecker
- Tables are not rendering, which means you did not review your changes as instructed in the Contribution guide you'll find in the README
- The very first TODO has a broken link

ikas commented:
@edwardpmorris 

Regarding the first issue you reported (error posting to /v1/geostore/area), it was in fact a bug, it will be fixed once https://github.com/gfw-api/gfw-geostore-api/pull/63 is merged and deployed - thanks for reporting :+1:

Regarding the second issue (`Requires authentication` when trying to view a geostore), that is not the Geostore service requiring authentication, but an error thrown by the Github service when trying to create a gist for that geojson (funny thing, you can get that geostore by hash, paste it in http://geojson.io/ and you get the exact same "Requires authentication" message :sweat:).

Not sure if we will be able to fix this (or even if it's worth fixing :man_shrugging:), but in any case:
* you can safely state in the docs that no endpoint from the Geostore service requires authentication. 
* for this endpoint in particular (`/v2/geostore/:hash/view`), add a note stating that for very large geometries (_i.e. geojson string has more than 150000 chars of length_) this endpoint won't work (i.e. it will return that cryptic "Requires authentication" message).

edwardpmorris commented:
Used spell checker, changed tables to long-style, removed TODO with link. Added info about `geostore/view`. I was unable to review the HTML rendering, had difficulty setting up the specific Ruby version.    

Outstanding issues:
+ Describe behavior when adding exact same GeoJSON. Is the geostore object overwritten, no action?
+ Provide an example of "Create using a provider definition" that works.
+ Give details about the WDPA version and update frequency (need to check with WRI).
+ Write section for `/v1/geostore/area` once bug fixed.

ikas reviewed (approved):
Added some minor commits, but overall looks great, good job @edwardpmorris :clap:

ikas commented:
Related to JIRA tickets [GTC-279](https://gfw.atlassian.net/browse/GTC-279) and [GTC-734](https://gfw.atlassian.net/browse/GTC-734).

ethanroday commented on source/includes/concept/_geostore.md:
```diff
@@ -0,0 +1,50 @@
+## Geostore
```
A few thoughts here:
 - The first two paragraphs are great - exactly what we need from concept documentation.
 - However, I think the rest of this section could be made quite a bit more readable by just being a little less formal 😄 For example, something more along the lines of: "The Resource Watch API allows users to define custom vector data structures in GeoJSON format. We call these custom areas "geostores". Each geostore consists of an id, some metadata, and the GeoJSON itself."
 - Also, I would replace "Geostore microservice" with "Geostore API", since that's what users are actually interacting with.
 - We need to devote some space here to talking about what the geostore API is actually used for (e.g. it's what underlies the areas service, etc.)

ikas replied:
Thanks for the feedback Ethan :bow: 

- Reworded the final paragraphs of the concept docs and tried to simplify the text a bit so that the overall tone is less formal.
- Replaced "Geostore microservice" with "Geostore API".
- Added a section "What are geostores used for?" at the end of the concept docs.

ethanroday commented on source/includes/concept/_geostore.md:
```diff
@@ -0,0 +1,50 @@
+- GeoJSON [only supports one geographic coordinate reference system [CRS]](https://tools.ietf.org/html/rfc7946#section-4), using the World Geodetic System 1984 (WGS 84) datum, with longitude and latitude units of decimal degrees. Other CRS are not supported. Geostore does not check for the validity of the coordinate CRS.
+- During creation the GeoJSON geometry is checked for geometric validity and, if required, an attempt is made to repair the geometry (using [ST_MakeValid](https://postgis.net/docs/ST_MakeValid.html)).
+
+### Geostore id and hash
```
I would say everything past this point can be included in the front matter of the endpoint documentation. No need to get this technical in the concept documentation.

ikas replied:
Removed this information from the concept docs, since it's already being covered at the start of the endpoint docs.

ethanroday commented on source/includes/api/_geostore.md:
```diff
@@ -1,92 +1,112 @@
+This identifier is a 128-bit hash generated by passing the [pre-processed GeoJSON string](#geostore) through the [MD5 message-digest algorithm](https://en.wikipedia.org/wiki/MD5). This means that each identifier is a unique fingerprint of the GeoJSON string.
 
-### With a Geojson
+So if you add an exact copy of your GeoJSON string it will generate the same hash, but if the GeoJSON string is changed in any way it will have a different hash. Hence, the Geostore identifier provides the RW API a way to ensure that platforms and microservices are using *exactly* the same geographic structures. *Note changes in the Geostore id may also represent changes in other  GeoJSON propert ...
```
If the id is generated from the GeoJSON, does this mean geostore objects are static and can't be modified? If so, I would point that out here.

ikas replied:
That's right, thanks for pointing out - I have reworded this paragraph and added the missing piece of information.

ethanroday commented on source/includes/api/_geostore.md:
```diff
@@ -1,92 +1,112 @@
+
+## Overview of available endpoints
+
+The RW APIs Geostore feature has a number of endpoints to help users interact with geostore objects. Many of them are essentially helpers to make it easy to access geometries from useful data sources, such as:
```
```suggestion
The Geostore API consists of a number of endpoints to help users interact with geostore objects. Many of them are essentially helpers to make it easy to access geometries from useful data sources, such as:
```

ethanroday commented on source/includes/api/_geostore.md:
```diff
@@ -125,395 +145,845 @@ curl -X POST https://api.resourcewatch.org/v1/geostore \
+### Get geostore by country code
 
-Geostore has the geojson of 4 different lands use:
+Geostore objects can be retrieved via the GET `geostore/admin/:iso` endpoint, which returns a single object selected from the Geostore using a single  country [ISO 3166-1 alpha-3 code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3) (`<iso>`). The returned geometries can be simplified using the parameter `?simplify=<value>`. Where `<value>` is given to the simplification algorithm, 0.005 usuall ...
```
Just to be clear, the geostore that's returned is the boundary of the given country, right? This wording really doesn't make that clear.

ikas replied:
Yes, that's right - I have reworded the initial paragraphs of all these endpoints (getting geostores by country code, ADM1, ADM2, WDPA, land use type) so they are a bit clearer about the returned data.

ethanroday commented on source/includes/api/_geostore.md:
```diff
@@ -125,395 +145,845 @@ curl -X POST https://api.resourcewatch.org/v1/geostore \
-### From land use areas
+*Note - This endpoint does not return a unique array of iso codes and names.*
+
+### Get geostore by country code
```
I would put all of these (country code, ADM1, ADM2, WDPA, land use type) in their own section called something like "Getting predefined geostores for commonly used geometries". Point out that this is essentially a repository of useful geometries that is provided as a utility function of sorts. The fact that they are stored with the same underlying mechanism as custom geostores is really just a technical matter.

tiagojsag commented on source/includes/api/_geostore.md:
````diff
@@ -125,395 +145,857 @@ curl -X POST https://api.resourcewatch.org/v1/geostore \
 
 ```shell
-curl -X GET https://api.resourcewatch.org/v1/geostore/use/oilpalm/<id>
+curl -X GET https://api.resourcewatch.org/v2/geostore/admin/<iso>simplify=<value>
````
this URL is wrong,please review the  simplify param

ikas replied:
Removed :+1:

tiagojsag commented on source/includes/api/_geostore.md:
```diff
@@ -1,92 +1,112 @@
+| POST   | `/v2/geostore/find-by-ids`          | [Get geostore objects by Geostore ids](#get-geostore-objects-by-geostore-ids) |
+| GET    | `/v2/geostore/:id/view`             | [Get a geostore object by Geostore id and view at GeoJSON.io](#get-a-geostore-object-by-geostore-id-and-view-at-geojson-io) |
+| GET    | `/v2/geostore/admin/list`           | [Get all Geostore ids, names and country codes](#get-all-geostore-ids-names-and-country-codes) |
+| GET    | `/v2/geostore/admin/:iso`           | [Get a geostore object by country code](#get-country-code)                    |
```
Cant seem to find docs for this particular endpoint

ikas replied:
Link was incorrect, fixed

tiagojsag reviewed (changes requested):
left a few notes based on me trying to look for info on a particular endpoint, and not getting the necessary info. I would say we may need to do a proper review at this, to make sure it's complete

ikas commented:
@tiagojsag all endpoints in the overview section have been covered, but the reference links were not correct, please check again if you can find the info you need.
